dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/utils/MemoizationUtil.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2022-2023 Google LLC
+ * Copyright 2013-2021 CompilerWorks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.edwmigration.dumper.application.dumper.utils;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.commons.io.function.IOFunction;
+
+/** Utility class for memoization. */
+public class MemoizationUtil {
+
+  private static class Memoizer<T, K, R> implements IOFunction<T, R> {
```

Comment:
I would recommend using a standard caching mechanism, e.g. guava `LoadingCache` with mapping the key to the `DriverInformation` in the loading function instead of creating a custom solution (`Memoizer` class).

The aliases could be moved from `driverInformationMap` to a new `aliasesMap`.

Reply:
I considered using `LoadingCache` but it is not a good fit. It is not really a cache and hence I don't need any of the eviction features. Handling of `IOException`s gets more complicated with `LoadCache`.

I don't strictly need the key handling I implemented (since the `DriverInformation` instances are all the same object for all aliases) but it is still nice that we can be sure that we point to the same URI.

The heavy lifting here is really done by `HashMap`. In that sense, what I need is a hash map, not a cache and `LoadingCache` would just save me from doing line 41-47 in `MemoizationUtil`.

Reply:
> It is not really a cache and hence I don't need any of the eviction features.

From Guava manual:
"you should consider using caches when a value is expensive to compute or retrieve, and you will need its value on a certain input more than once. (...) In some cases a `LoadingCache` can be useful even if it doesn't evict entries, due to its automatic cache loading."
https://github.com/google/guava/wiki/CachesExplained#applicability

> I don't strictly need the key handling I implemented (since the `DriverInformation` instances are all the same object for all aliases) but it is still nice that we can be sure that we point to the same URI.

Yes, I would slightly suggest to either map URIs to the Path of the downloaded driver, or use the whole `DriverInformation` as a key. In both approaches the `Memoizer` could be simplified to require 2 generic parameters instead of 3.

> `LoadingCache` would just save me from doing line 41-47 in `MemoizationUtil`

This is my main point - to reuse existing code that was written for this exact purpose (unless you think using `LoadingCache` is an overkill in this case).

Reply:
The issue with `IOException` remains. The handling of it in `LoadingCache` introduces more code than what `MemoizationUtil` does. Otherwise, I would use `LoadingCache`.

Reply:
To explore this a bit further I created https://github.com/google/dwh-migration-tools/compare/feature-cache-dependency-downloads...feature-cache-dependency-downloads-with-loading-cache?expand=1

As you can see, the exception handling adds more code. But not just that, imagine debugging an issue: You know have to understand `LoadingCache`. Most Java developers have a decent understanding of maps but will not know the details of Guava's caching. So, by using `LoadingCache` it will be harder to reason about what is happening.

Reply:
The code:
```
public static <T, R> IOFunction<T, R> createMemoizer(IOFunction<T, R> delegate) {
  LoadingCache<T, R> cache =
      CacheBuilder.newBuilder()
          .build(
              new CacheLoader<T, R>() {
                public R load(T t) throws IOException {
                  return delegate.apply(t);
                }
              });
  return t -> {
    try {
      return cache.get(t);
    } catch (UncheckedExecutionException e) {
      if (e instanceof RuntimeException) {
        throw (RuntimeException) e.getCause();
      } else {
        throw new RuntimeException("Got unexpected UncheckedExecutionException.", e);
      }
    } catch (ExecutionException e) {
      Throwable cause = e.getCause();
      if (cause instanceof IOException) {
        throw (IOException) cause;
      } else {
        throw new RuntimeException("Got unexpected ExecutionException.", e);
      }
    }
  };
}
```

could be simplified to:

```
public static <T, R> Function<T, R> memoize(Function<T, R> delegate) {
  LoadingCache<T, R> cache = CacheBuilder.newBuilder().build(CacheLoader.from(delegate));
  return t -> {
    try {
      return cache.get(t);
    } catch (ExecutionException e) {
      throw new RuntimeException("Got unexpected ExecutionException.", e);
    }
  };
}
```
The unwrapping of some exceptions, yet keeping others wrapped, might not bring a lot of simplification compared to the shorter version. However, the shorter version would require handling of `IOException` either in the Memoizer or inside of delegate (e.g. converting it to `UncheckedIOException`). Exceptions and lambdas require choosing some tradeoff.

If you want to use `IOFunction` and `IOException`, then it is probably not possible to write a much simpler code using `LoadingCache` than the exploration you've prepared. In this case using `HashMap` makes more sense.